arch/x86/core/x86_mmu.c

```diff
@@ -97,53 +40,29 @@ void _x86_mmu_get_flags(void *addr, u32_t *pde_flags, u32_t *pte_flags)
 }
 
 
-/**
- * @brief check page table entry flags
- *
- * This routine checks if the buffer is avaialable to the whoever calls
- * this API.
- * @a addr start address of the buffer
- * @a size the size of the buffer
- * @a flags permisions to check.
- *    Consists of 2 bits the bit0 represents the RW permissions
- *    The bit1 represents the user/supervisor permissions
- *    Use macro BUFF_READABLE/BUFF_WRITEABLE or BUFF_USER to build the flags
- *
- * @return true-if the permissions of the pde matches the request
- */
-int _x86_mmu_buffer_validate(void *addr, size_t size, int flags)
+int _arch_buffer_validate(void *addr, size_t size, int write)
 {
-	int status = 0;
 	u32_t start_pde_num;
 	u32_t end_pde_num;
-	/* union x86_mmu_pde_pt pde_status; */
-	volatile u32_t validity = 1;
 	u32_t starting_pte_num;
 	u32_t ending_pte_num;
 	struct x86_mmu_page_table *pte_address;
-	u32_t rw_permission;
-	u32_t us_permission;
 	u32_t pde;
 	u32_t pte;
 	union x86_mmu_pte pte_value;
 
 	start_pde_num = MMU_PDE_NUM(addr);
 	end_pde_num = MMU_PDE_NUM((char *)addr + size - 1);
-	rw_permission = GET_RW_PERM(flags);
-	us_permission = GET_US_PERM(flags);
 	starting_pte_num = MMU_PAGE_NUM((char *)addr);
 
 	/* Iterate for all the pde's the buffer might take up.
 	 * (depends on the size of the buffer and start address of the buff)
 	 */
 	for (pde = start_pde_num; pde <= end_pde_num; pde++) {
-		validity &= check_pde_flags(X86_MMU_PD->entry[pde].pt,
-					   rw_permission,
-					   us_permission);
+		union x86_mmu_pde_pt pde_value = X86_MMU_PD->entry[pde].pt;
 
-		/* If pde is invalid exit immediately. */
-		if (validity == 0) {
-			break;
+		if (!pde_value.p || !pde_value.us || (write && !pde_value.rw)) {
+			return 0;
```

Comment:
AFAIK we return a non-zero value to denote a failure.
If we need a true/false type of output. Lets make this API a bool instead of a int.
Because this is very confusing.

Reply:
This is what was in the docstring:
"@return true-if the permissions of the pde matches the request"

I am returning 0 (false) if the permissions do not match, and nonzero (true) if they do match. Given how C handles true and false, this is what is expected and documented. The code was the opposite of the documentation.